src/Ship.cpp

```diff
@@ -897,13 +897,20 @@ void Ship::Blastoff()
 
 	assert(f->GetBody()->IsType(ObjectType::PLANET));
 
-	const double planetRadius = 2.0 + static_cast<Planet *>(f->GetBody())->GetTerrainHeight(up);
-	SetVelocity(vector3d(0, 0, 0));
-	SetAngVelocity(vector3d(0, 0, 0));
-	SetFlightState(FLYING);
+	if (ManualDocking()) {
+		// no one except the player can take off without the autopilot
+		assert(IsType(ObjectType::PLAYER));
```

Comment:
I'd strongly prefer this be a runtime check with a soft-fail condition rather than an assertion. If the precondition is that this branch can only be taken by an object of type Player, then the logic should live in a method on Player instead - from a separation-of-concerns perspective at least.

Reply:
I just give a warning and return, is it ok? Or is more serious error handling required?

Reply:
That's fine, I'm mostly concerned with it being a "recoverable" error rather than an immediate crash.